File: frontend/src/api/wiki.ts
import { api } from './client'

export interface WikiPage {
  id: number
  path: string
  title: string
  page_type: string
  page_key?: string | null
  domain?: string | null
  status?: string | null
  revision?: number | null
  aliases?: string[]
  source_document_id: number | null
  tags: string[]
  content: string | null
  created_at: string
  updated_at: string
}

export interface WikiIndex {
  content: string
  path: string
}

export interface RetrieveHit {
  id: number | null
  title: string
  page_type: string
  path: string
  page_key?: string | null
  domain?: string | null
  status?: string | null
  revision?: number | null
  score: number
  snippet: string
  tags: string[]
  content: string | null
  source_document_id: number | null
  /** wiki | source */
  citation_type?: string
  source_chunk_id?: number | null
  start_char?: number | null
  end_char?: number | null
  clause_ids?: string[]
  anchor_clause?: string | null
  explain?: Record<string, unknown> | null
  aliases?: string[]
  source_document_ids?: number[]
}

export interface RetrieveResponse {
  query: string
  hits: RetrieveHit[]
  wiki_hit_count?: number
  source_hit_count?: number
  clause_ids?: string[]
  anchored_clause_ids?: string[]
  retrieval_mode?: string | null
  explain?: Record<string, unknown> | null
}

export type WikiReviewStatus = 'pending' | 'approved' | 'rejected' | 'acknowledged'

export interface WikiReview {
  id: number
  page_id: number | null
  job_id: number | null
  kind: string
  status: WikiReviewStatus | string
  reason: string
  candidate_available: boolean
  reviewed_at: string | null
  reviewed_by: string | null
  decision_reason: string | null
  created_at: string
  updated_at: string
}

export interface WikiSourceEvidence {
  document_id: number
  chunk_ids: number[]
  clauses: string[]
}

export interface WikiRevision {
  id: number
  page_id: number
  revision: number
  frontmatter: Record<string, unknown>
  frontmatter_json: string
  content_md: string
  operation: string
  job_id: number | null
  reason: string
  created_at: string
}

export interface WikiCandidate {
  page_key: string | null
  title: string | null
  type: string | null
  domain: string | null
  aliases: string[]
  tags: string[]
  status: string | null
  sources: WikiSourceEvidence[]
  content_md: string | null
  raw: Record<string, unknown>
}

export interface WikiReviewReason {
  summary: string
  kind: string
  operation: string | null
  page_key: string | null
  risk_flags: string[]
}

export interface WikiDiff {
  from_revision: number | null
  to_revision: number | null
  unified: string
  text: string
  changed: boolean
  available: boolean
  reason: string
}

export interface WikiReviewDetail extends WikiReview {
  old_version: WikiRevision | null
  new_candidate: WikiCandidate
  reason_detail: WikiReviewReason
  payload: Record<string, unknown>
  source_evidence: WikiSourceEvidence[]
  diff: WikiDiff
}

export interface WikiReviewDecision {
  reviewed_by?: string
  reason?: string
  decision_reason?: string
}

export interface WikiRollbackRequest {
  revision_id?: number
  revision?: number
  job_id?: number
  reason: string
  reviewed_by?: string
}

export type WikiRollbackResponse = WikiRevision

export interface SourceChunk {
  id: number
  document_id: number
  chunk_index: number
  title: string
  text: string
  start_char: number
  end_char: number
}

export function listWikiPages() {
  return api<WikiPage[]>('/api/wiki/pages')
}

export function getWikiPage(id: number) {
  return api<WikiPage>(`/api/wiki/pages/${id}`)
}

export function getSourceChunk(id: number) {
  return api<SourceChunk>(`/api/source-chunks/${id}`)
}

export function getWikiIndex() {
  return api<WikiIndex>('/api/wiki/index')
}

export function retrieveWiki(query: string, top_k?: number) {
  return api<RetrieveResponse>('/api/wiki/retrieve', {
    method: 'POST',
    body: JSON.stringify({ query, top_k }),
  })
}

export interface WikiReviewQuery {
  status?: WikiReviewStatus | string
  kind?: string
  page_id?: number
  job_id?: number
}

function withQuery(path: string, params: Record<string, string | number | undefined>) {
  const search = new URLSearchParams()
  Object.entries(params).forEach(([key, value]) => {
    if (value !== undefined && value !== null && value !== '') {
      search.set(key, String(value))
    }
  })
  const query = search.toString()
  return query ? `${path}?${query}` : path
}

export function listWikiReviews(params: WikiReviewQuery = {}) {
  return api<WikiReview[]>(
    withQuery('/api/wiki/reviews', {
      status: params.status,
      kind: params.kind,
      page_id: params.page_id,
      job_id: params.job_id,
    }),
  )
}

export function getWikiReview(id: number) {
  return api<WikiReviewDetail>(`/api/wiki/reviews/${id}`)
}

export function approveWikiReview(id: number, body: WikiReviewDecision = {}) {
  return api<WikiReviewDetail>(`/api/wiki/reviews/${id}/approve`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function rejectWikiReview(id: number, body: WikiReviewDecision = {}) {
  return api<WikiReviewDetail>(`/api/wiki/reviews/${id}/reject`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function acknowledgeWikiReview(id: number, body: WikiReviewDecision = {}) {
  return api<WikiReviewDetail>(`/api/wiki/reviews/${id}/acknowledge`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

export function listWikiRevisions(pageId: number) {
  return api<WikiRevision[]>(`/api/wiki/pages/${pageId}/revisions`)
}

export function getWikiRevision(pageId: number, revisionId: number) {
  return api<WikiRevision>(`/api/wiki/pages/${pageId}/revisions/${revisionId}`)
}

export interface WikiDiffQuery {
  from_revision?: number
  to_revision?: number
}

export function getWikiDiff(pageId: number, params: WikiDiffQuery = {}) {
  return api<WikiDiff>(
    withQuery(`/api/wiki/pages/${pageId}/diff`, {
      from_revision: params.from_revision,
      to_revision: params.to_revision,
    }),
  )
}

export function rollbackWikiPage(pageId: number, body: WikiRollbackRequest) {
  return api<WikiRollbackResponse>(`/api/wiki/pages/${pageId}/rollback`, {
    method: 'POST',
    body: JSON.stringify(body),
  })
}

File: backend/app/schemas/wiki.py
from datetime import datetime
from typing import Any, List, Optional

from pydantic import BaseModel, ConfigDict, Field, model_validator


class IngestJobOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    document_id: int
    space_id: int
    space_name: str = ""
    status: str
    stage: str = "queued"
    progress: int = 0
    plan_json: str = "{}"
    cancel_requested: bool = False
    model_ref: Optional[str] = None
    prompt_version_ref: Optional[str] = None
    step_log_json: str
    error_message: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class WikiPageOut(BaseModel):
    model_config = ConfigDict(from_attributes=True)

    id: int
    path: str
    space_id: int
    space_name: str = ""
    title: str
    page_type: str
    source_document_id: Optional[int] = None
    page_key: Optional[str] = None
    domain: Optional[str] = None
    status: str = "published"
    revision: int = 1
    aliases: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    content: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class WikiIndexOut(BaseModel):
    content: str
    path: str = "wiki/index.md"


class RetrieveRequest(BaseModel):
    query: str
    space_id: Optional[int] = Field(default=None, ge=1)
    top_k: Optional[int] = None
    types: Optional[List[str]] = None


class RetrieveHit(BaseModel):
    id: Optional[int] = None
    title: str
    page_type: str
    path: str
    score: float
    snippet: str
    tags: List[str] = Field(default_factory=list)
    content: Optional[str] = None
    source_document_id: Optional[int] = None
    # wiki | source
    citation_type: str = "wiki"
    source_chunk_id: Optional[int] = None
    start_char: Optional[int] = None
    end_char: Optional[int] = None
    clause_ids: List[str] = Field(default_factory=list)
    anchor_clause: Optional[str] = None
    explain: Optional[dict] = None
    page_key: Optional[str] = None
    domain: Optional[str] = None
    status: Optional[str] = None
    revision: Optional[int] = None
    aliases: List[str] = Field(default_factory=list)
    source_document_ids: List[int] = Field(default_factory=list)
    space_id: Optional[int] = None


class RetrieveResponse(BaseModel):
    query: str
    hits: List[RetrieveHit]
    wiki_hit_count: int = 0
    source_hit_count: int = 0
    clause_ids: List[str] = Field(default_factory=list)
    anchored_clause_ids: List[str] = Field(default_factory=list)
    retrieval_mode: Optional[str] = None
    explain: Optional[dict] = None


class WikiSourceEvidenceOut(BaseModel):
    document_id: int
    chunk_ids: List[int] = Field(default_factory=list)
    clauses: List[str] = Field(default_factory=list)


class WikiDiffOut(BaseModel):
    from_revision: Optional[int] = None
    to_revision: Optional[int] = None
    unified: str = ""
    # ``text`` is a readable alias for clients that do not need to know the
    # diff format; ``unified`` remains the canonical field.
    text: str = ""
    changed: bool = False
    available: bool = True
    reason: str = ""


class WikiRevisionOut(BaseModel):
    id: int
    page_id: int
    revision: int
    frontmatter: dict[str, Any] = Field(default_factory=dict)
    frontmatter_json: str = "{}"
    content_md: str = ""
    operation: str = "update"
    job_id: Optional[int] = None
    reason: str = ""
    created_at: datetime


class WikiReviewOut(BaseModel):
    id: int
    page_id: Optional[int] = None
    job_id: Optional[int] = None
    space_id: Optional[int] = None
    space_name: str = ""
    kind: str
    status: str
    reason: str
    candidate_available: bool = False
    reviewed_at: Optional[datetime] = None
    reviewed_by: Optional[str] = None
    decision_reason: Optional[str] = None
    created_at: datetime
    updated_at: datetime


class WikiReviewReasonOut(BaseModel):
    summary: str = ""
    kind: str = ""
    operation: Optional[str] = None
    page_key: Optional[str] = None
    risk_flags: List[str] = Field(default_factory=list)


class WikiCandidateOut(BaseModel):
    page_key: Optional[str] = None
    title: Optional[str] = None
    type: Optional[str] = None
    domain: Optional[str] = None
    aliases: List[str] = Field(default_factory=list)
    tags: List[str] = Field(default_factory=list)
    status: Optional[str] = None
    sources: List[WikiSourceEvidenceOut] = Field(default_factory=list)
    content_md: Optional[str] = None
    raw: dict[str, Any] = Field(default_factory=dict)


class WikiReviewDetailOut(WikiReviewOut):
    old_version: Optional[WikiRevisionOut] = None
    new_candidate: WikiCandidateOut = Field(default_factory=WikiCandidateOut)
    reason_detail: WikiReviewReasonOut = Field(default_factory=WikiReviewReasonOut)
    payload: dict[str, Any] = Field(default_factory=dict)
    source_evidence: List[WikiSourceEvidenceOut] = Field(default_factory=list)
    diff: WikiDiffOut = Field(default_factory=WikiDiffOut)


class WikiReviewDecisionIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    reviewed_by: Optional[str] = None
    reason: Optional[str] = None
    decision_reason: Optional[str] = None


class WikiRollbackIn(BaseModel):
    model_config = ConfigDict(extra="forbid")

    revision_id: Optional[int] = Field(default=None, ge=1)
    revision: Optional[int] = Field(default=None, ge=1)
    job_id: Optional[int] = Field(default=None, ge=1)
    reason: str = Field(default="rollback", min_length=1)
    reviewed_by: Optional[str] = None

    @model_validator(mode="after")
    def _exactly_one_target(self) -> "WikiRollbackIn":
        if (self.revision_id is None) == (self.revision is None):
            raise ValueError("provide exactly one of revision_id or revision")
        return self


class WikiRollbackOut(WikiRevisionOut):
    pass
